crmsh/upgradeutil.py

```diff
@@ -0,0 +1,112 @@
+
+
+def _upgrade_if_needed(nodes, fn: typing.Callable[[int], None]):
+    seq = 0
+    forced = 0 == subprocess.call(['test', '-f', FORCE_UPGRADE_FILE_PATH])
+    needed = forced
+    if not needed:
+        try:
+            local_seq = int(_get_file_content(SEQ_FILE_PATH, b'').strip())
+        except ValueError:
+            local_seq = 0
+        needed = local_seq < CURRENT_UPGRADE_SEQ
+    if not needed:
+        results = list(_parallax_run(nodes, 'cat {}'.format(SEQ_FILE_PATH)).values())
+        try:
+            seq = min(int(stdout.strip()) if rc == 0 else 0 for rc, stdout, stderr in results)
+        except ValueError:
+            seq = 0
+        needed = seq < CURRENT_UPGRADE_SEQ
+    if needed:
+        try:
+            fn(seq)
+        except _SkipUpgrade:
+            logger.warning("Upgrade skipped.")
+            return
+        crmparallax.parallax_call(
+            nodes,
+            "mkdir -p '{data_dir}' && echo '{seq}' > '{seq_file_path}'".format(
+                data_dir=DATA_DIR, seq=seq, seq_file_path=SEQ_FILE_PATH,
+            ),
+        )
+        crmparallax.parallax_call(nodes, "echo '{}' > '{}'".format(CURRENT_UPGRADE_SEQ, SEQ_FILE_PATH))
+        crmparallax.parallax_call(nodes, 'rm -f {}'.format(FORCE_UPGRADE_FILE_PATH))
+
+
+def _upgrade(nodes, seq):
+    logger.info("Upgrading cluster...")
+    if seq == 0:
+        seq_0_setup_hacluster_passwordless(nodes)
+
+
+def upgrade_if_needed():
+    nodes = crmutils.list_cluster_nodes()
+    if nodes:
+        _upgrade_if_needed(nodes, lambda seq: _upgrade(nodes, seq))
+
+
+def force_set_local_upgrade_seq():
+    try:
+        os.mkdir(DATA_DIR)
+    except FileExistsError:
+        pass
+    with open(SEQ_FILE_PATH, 'w', encoding='ascii') as f:
+        print(CURRENT_UPGRADE_SEQ, file=f)
+
+
+def seq_0_setup_hacluster_passwordless(nodes):
+    results = _parallax_run(nodes, '[ -f ~hacluster/.ssh/id_rsa ] || [ -f ~hacluster/.ssh/id_ecdsa ] || [ -f ~hacluster/.ssh/id_ed25519 ]')
+    for node, result in results.items():
+        if isinstance(result, Exception):
+            logger.error('Failed to connect host %s.', node, exc_info=result)
+            raise result
+    nodes_without_keys = [node for node, result in results.items() if result[0] != 0]
+    if nodes_without_keys:
+        if not crmutils.ask("Upgrade: setup passwordless ssh authentication for user hacluster?"):
+            raise _SkipUpgrade()
+        init_node = nodes_without_keys[0]
+        join_nodes = list()
+        join_nodes.extend(nodes)
+        join_nodes.remove(init_node)
+        crmparallax.parallax_call([init_node], 'crm cluster init ssh -y')
+        for node in join_nodes:
+            result = next(iter(_parallax_run([node], 'crm cluster join ssh -c {} -y'.format(init_node)).values()))
+            if isinstance(result, Exception):
+                logger.error('Failed to join ssh on host %s.', node, exc_info=result)
+                raise result
```

Comment:
I think It'd be better to add one or two log to describe what is this process doing:)